### PR TITLE
[Arctic-44][Hive][AMS]Implement util to upgade hive table to arctic table with hive base store

### PR DESCRIPTION
**Pull request by shendanfengg**
## Why are the changes needed?

*Arctic Planned in 0.3.1 supports Hive table compatibility.  AMS Dash Dashboard need to support for upgrade a Hive table  to an Arctic table and also hive moudle need to move Hive data file to Arctic base table.*

## Brief change log

  - *Support ams upgrade a Hive table to Arctic table*
  - *Support migrate Hive data file to Arctic Base table*

## How was this patch tested?
An environment was constructed and four types of tables were tested include pk partition table, no pk partition table, pk no partition table, no pk no partition table 

## Documentation

  - This pull request introduces a new feature.
image
image
image

### CODE REVIEW

**Review by zhoujinsong — CHANGES_REQUESTED** [review]

**Comment by SteNicholas** [comment]
@shendanfengg, could you please update the title from `impelement` to `implement`?

File: hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java [hunk 1]
```diff
@@ -173,12 +191,15 @@ public ArcticHiveTableBuilder(TableIdentifier identifier, Schema schema) {
     protected void doCreateCheck() {
       super.doCreateCheck();
       try {
-        org.apache.hadoop.hive.metastore.api.Table hiveTable =
-            hiveClientPool.run(client -> client.getTable(
-                identifier.getDatabase(),
-                identifier.getTableName()));
-        if (hiveTable != null) {
-          throw new IllegalArgumentException("Table is already existed in hive meta store:" + identifier);
+        if (!(properties.containsKey(TableProperties.UPGRADE_ENABLE) &&
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java]: It has been modified.

File: ams/ams-server/pom.xml [hunk 2]
```diff
@@ -74,24 +74,26 @@
...
-<!--                    <artifactId>hadoop-yarn-api</artifactId>-->
-<!--                </exclusion>-->
-<!--                <exclusion>-->
-<!--                    <artifactId>libthrift</artifactId>-->
-<!--                    <groupId>org.apache.thrift</groupId>-->
-<!--                </exclusion>-->
-<!--            </exclusions>-->
-<!--        </dependency>-->
+
+        <dependency>
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply ams/ams-server/pom.xml:78]: It has been deleted.

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/config/ServerTableProperties.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.netease.arctic.table.TableProperties;
+
+import java.awt.event.HierarchyBoundsAdapter;
+import java.util.HashSet;
+import java.util.Set;
+
+public class ServerTableProperties {
+
+  public static final Set<String> HIDDEN_EXPOSED = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/config/ServerTableProperties.java:29]: Will commit in a new issue

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/controller/TableController.java [hunk 4]
```diff
@@ -143,19 +157,92 @@ public static void getTableDetail(Context ctx) {
...
+
+    // get table from catalog
+    String thriftHost = ArcticMetaStore.conf.getString(ArcticMetaStoreConf.THRIFT_BIND_HOST);
+    Integer thriftPort = ArcticMetaStore.conf.getInteger(ArcticMetaStoreConf.THRIFT_BIND_PORT);
+    ArcticHiveCatalog ac = (ArcticHiveCatalog)CatalogUtil.getArcticCatalog(thriftHost, thriftPort, catalog);
+
+    TableIdentifier tableIdentifier = TableIdentifier.of(catalog, db, table);
+    HiveTableInfo hiveTableInfo = null;
+    try {
+      HiveMetaStore hiveMetaStore = HiveMetaStore.getHiveMetaStore(ac);
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/controller/TableController.java]: It has been modified

File: ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/MetaService.java [hunk 5]
```diff
@@ -58,6 +58,7 @@ public static ServerTableMeta getServerTableMeta(ArcticCatalog ac, TableIdentifi
             TableProperties.TABLE_CREATE_TIME_DEFAULT));
 
     TableProperties.PROTECTED_PROPERTIES.forEach(serverTableMeta.getProperties()::remove);
+    serverTableMeta.getProperties().remove(TableProperties.TABLE_CREATE_TIME);
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply ams/ams-server/src/main/java/com/netease/arctic/ams/server/service/MetaService.java:62]: This propertie is used for frontend.

File: core/src/main/java/com/netease/arctic/io/ArcticFileIO.java [hunk 6]
```diff
@@ -46,6 +46,23 @@ public interface ArcticFileIO extends FileIO {
...
+  /**
+   * Create a new directory.
+   *
+   * @param path source path
+   * @return true if the create success;
+   */
+  boolean mkdirs(String path);
+
+  /**
+   * Create a new directory.
```
Comment (zhoujinsong) [inline core/src/main/java/com/netease/arctic/io/ArcticFileIO.java:58]: Wrong javadoc

File: hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java [hunk 7]
```diff
@@ -166,22 +180,30 @@ protected UnkeyedHiveTable loadUnKeyedTable(TableMeta tableMeta) {
...
     protected void doCreateCheck() {
       super.doCreateCheck();
       try {
-        org.apache.hadoop.hive.metastore.api.Table hiveTable =
-            hiveClientPool.run(client -> client.getTable(
-                identifier.getDatabase(),
-                identifier.getTableName()));
-        if (hiveTable != null) {
-          throw new IllegalArgumentException("Table is already existed in hive meta store:" + identifier);
+        if (!isUpgradeHive) {
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java]: I recommend to add a `allowHiveTableExsited` config for table builder and check existed hive table schema and partition spec.

File: hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java [hunk 8]
```diff
@@ -331,27 +373,47 @@ protected String getDatabaseLocation() {
...
-        LOG.warn("Failed to drop hive table while rolling back create table operation", e);
+      if (isUpgradeHive) {
+        LOG.info("No need to drop hive table");
+        com.netease.arctic.ams.api.TableIdentifier tableIdentifier = meta.getTableIdentifier();
+        try {
+          hiveClientPool.run(client -> {
+            org.apache.hadoop.hive.metastore.api.Table hiveTable = client.getTable(tableIdentifier.getDatabase(),
+                tableIdentifier.getTableName());
+            Map<String, String> hiveParameters = hiveTable.getParameters();
+            hiveParameters.put(HiveTableProperties.ARCTIC_TABLE_FLAG, "false");
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java:391]: I recommend to delete this properties instaned of set it value to false.

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.List;
+import java.util.UUID;
+
+public class HiveMigrateUtil {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HiveMigrateUtil.class);
+
+  private static final String DEFAULT_TXID = "txid=0";
+
+  public static void hiveDataMigration(ArcticTable arcticTable, ArcticHiveCatalog ac, TableIdentifier tableIdentifier)
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java:39]: Not use short name `ac` in method definition

Comment (zhoujinsong) [reply hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java:39]: tableIdentifier would be ge by arcticTable?

Comment (zhoujinsong) [reply hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java:39]: add java doc for public static method

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+      throws Exception {
+    Table hiveTable = HiveTableUtil.loadHmsTable(ac.getHMSClient(), tableIdentifier);
+    String hiveDataLocation = hiveTable.getSd().getLocation() + "/hive";
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java:42]: Use `HiveTableUtil.hiveRootLocation`

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    arcticTable.io().mkdirs(hiveDataLocation);
+    String newPath;
+    if (hiveTable.getPartitionKeys().isEmpty()) {
+      newPath = hiveDataLocation + "/" + System.currentTimeMillis() + "_" + UUID.randomUUID();
+      arcticTable.io().mkdirs(newPath);
+      for (FileStatus fileStatus : arcticTable.io().list(hiveTable.getSd().getLocation())) {
+        if (hiveTable.getPartitionKeys().isEmpty()) {
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java:49]: duplicate if statement with line 45

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveSchemaUtil.java [hunk 12]
```diff
@@ -33,6 +30,8 @@
  */
 public class HiveSchemaUtil {
 
+  private static final int PARTITION_COLUMN_START_ID = 2;
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveSchemaUtil.java:33]: not useable value

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java [hunk 13]
```diff
@@ -134,4 +143,138 @@ public static StorageDescriptor storageDescriptor(
     storageDescriptor.setSerdeInfo(serDeInfo);
     return storageDescriptor;
   }
+
+  public List<Partition> getHiveAllPartitions(HMSClient hiveClient, TableIdentifier tableIdentifier) {
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java:147]: I recommend to move it to HivePartitionUtil

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java [hunk 14]
```diff
@@ -134,4 +143,138 @@ public static StorageDescriptor storageDescriptor(
...
+      throw new NoSuchTableException(e, "Hive table does not exist: %s", tableIdentifier.getTableName());
+    } catch (TException e) {
+      throw new RuntimeException("Failed to get partitions " + tableIdentifier.getTableName(), e);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted in call to listPartitions", e);
+    }
+  }
+
+  public static List<String> getHivePartitionNames(HMSClient hiveClient, TableIdentifier tableIdentifier) {
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java:161]: recommend to move it to HivePartitionUtil

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java [hunk 15]
```diff
@@ -134,4 +143,138 @@ public static StorageDescriptor storageDescriptor(
...
+      throw new NoSuchTableException(e, "Hive table does not exist: %s", tableIdentifier.getTableName());
+    } catch (TException e) {
+      throw new RuntimeException("Failed to get partitions " + tableIdentifier.getTableName(), e);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted in call to listPartitions", e);
+    }
+  }
+
+  public static List<String> getHivePartitionLocations(HMSClient hiveClient,
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java:176]: recommend to move it to HivePartitionUtil

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java [hunk 16]
```diff
@@ -134,4 +143,138 @@ public static StorageDescriptor storageDescriptor(
...
+      throw new NoSuchTableException(e, "Hive table does not exist: %s", tableIdentifier.getTableName());
+    } catch (TException e) {
+      throw new RuntimeException("Failed to get partitions " + tableIdentifier.getTableName(), e);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted in call to listPartitions", e);
+    }
+  }
+
+  public static void alterPartition(HMSClient hiveClient, TableIdentifier tableIdentifier,
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java:195]: recommend to move it to HivePartitionUtil and consider resue `HivePartitionUtil.rewriteHivePartitions`

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java [hunk 17]
```diff
@@ -134,4 +143,138 @@ public static StorageDescriptor storageDescriptor(
...
+    } catch (TException e) {
+      throw new RuntimeException("Failed to get tables of database " + database, e);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted in call to getAllTables", e);
+    }
+    throw new RuntimeException("Failed to get tables of database " + database);
+  }
+
+  private static void reGetAllHiveTables(HMSClient hiveClient, MetaException e, String database) {
```
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java:248]: It seems not be used.

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java (hunk 9 above)
Comment (zhoujinsong) [inline hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java:39]: Add createing arctic table logic here too and change method name to `upgradeHiveTableToArctic`

**Review by zhoujinsong — CHANGES_REQUESTED** [review]

File: hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java [hunk 18]
```diff
@@ -112,16 +111,35 @@ public void dropDatabase(String databaseName) {
...
-      hiveClientPool.run(client -> {
-        client.dropTable(meta.getTableIdentifier().getDatabase(),
-            meta.getTableIdentifier().getTableName(),
-            purge /* deleteData */,
-            false /* ignoreUnknownTab */);
-        return null;
-      });
-    } catch (TException | InterruptedException e) {
-      throw new RuntimeException("Failed to drop table:" + meta.getTableIdentifier(), e);
+    if (meta.properties.containsKey(TableProperties.UPGRADE_ENABLE) &&
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java]: It has been modified

File: ams/ams-server/pom.xml [hunk 19]
```diff
@@ -152,6 +147,12 @@
             </exclusions>
         </dependency>
 
+        <dependency>
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply ams/ams-server/pom.xml]: It has been deleted

File: hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java [hunk 20]
```diff
@@ -166,22 +180,30 @@ protected UnkeyedHiveTable loadUnKeyedTable(TableMeta tableMeta) {
...
+    return hiveClientPool;
+  }
+
   class ArcticHiveTableBuilder extends BaseArcticTableBuilder {
 
     public ArcticHiveTableBuilder(TableIdentifier identifier, Schema schema) {
       super(identifier, schema);
     }
 
+
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java]: It has been deleted

File: core/src/main/java/com/netease/arctic/table/TableBuilder.java [hunk 21]
```diff
@@ -69,6 +69,14 @@ public interface TableBuilder {
    */
   TableBuilder withPrimaryKeySpec(PrimaryKeySpec primaryKeySpec);
 
+  /**
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply core/src/main/java/com/netease/arctic/table/TableBuilder.java]: It has been moved to table propeties

File: core/src/main/java/com/netease/arctic/io/ArcticFileIO.java (hunk 6 above)
Comment (shendanfengg) [reply core/src/main/java/com/netease/arctic/io/ArcticFileIO.java]: It has been modified

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java (hunk 14 above)
Comment (shendanfengg) [reply hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java]: It has been moved

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java (hunk 16 above)
Comment (shendanfengg) [reply hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java]: It has been moved but cannot reuse HivePartitionUtil.rewriteHivePartitions

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java (hunk 17 above)
Comment (shendanfengg) [reply hive/src/main/java/com/netease/arctic/hive/utils/HiveTableUtil.java]: It has been deleted

File: hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java (hunk 9 above)
Comment (shendanfengg) [reply hive/src/main/java/com/netease/arctic/hive/utils/HiveMigrateUtil.java]: It has been renamed

**Review by zhoujinsong — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/netease/arctic/table/TableProperties.java [hunk 22]
```diff
@@ -52,6 +51,7 @@ private TableProperties() {
 
   public static final String TABLE_EVENT_TIME_FIELD = "table.event-time-field";
 
+  public static final String ALLOW_HIVE_TABLE_EXISTED = "allow.hive-table.existed";
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply core/src/main/java/com/netease/arctic/table/TableProperties.java]: It has been moved

File: hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java [hunk 23]
```diff
@@ -176,12 +193,17 @@ public ArcticHiveTableBuilder(TableIdentifier identifier, Schema schema) {
     protected void doCreateCheck() {
       super.doCreateCheck();
       try {
-        org.apache.hadoop.hive.metastore.api.Table hiveTable =
-            hiveClientPool.run(client -> client.getTable(
-                identifier.getDatabase(),
-                identifier.getTableName()));
-        if (hiveTable != null) {
-          throw new IllegalArgumentException("Table is already existed in hive meta store:" + identifier);
+        if ((properties.containsKey(TableProperties.ALLOW_HIVE_TABLE_EXISTED) &&
```
[the remark this answers is not in the document]
Comment (shendanfengg) [reply hive/src/main/java/com/netease/arctic/hive/catalog/ArcticHiveCatalog.java]: It has been modified

**Review by zhoujinsong — APPROVED** [review]
Thanks for your great work.

**Pull request merged**